packages/embark-web3/src/index.js

```diff
@@ -0,0 +1,123 @@
+/* global __dirname module process require */
+
+const { dappPath, embarkPath, normalizePath, toForwardSlashes } = require('embark-utils');
+const constants = require('embark-core/constants');
+const path = require('path');
+
+class EmbarkWeb3 {
+  constructor(embark, _options) {
+    this.embark = embark;
+    this.logger = embark.logger;
+    this.events = embark.events;
+    this.fs = embark.fs;
+    this.config = embark.config;
+    this.modulesPath = dappPath(embark.config.embarkConfig.generationDir, constants.dappArtifacts.symlinkDir);
+
+    this.addWeb3ToEmbarkJS();
+  }
+
+  async addWeb3ToEmbarkJS() {
+    let blockchainConnectorReady = false;
+    await this.whenRuncodeReady();
+
+    const web3LocationPromise = this.getWeb3Location();
+
+    this.events.setCommandHandler('blockchain:connector:ready', (cb) => {
+      if (blockchainConnectorReady) {
+        return cb();
+      }
+      this.events.once("blockchain:connector:ready", () => {
+        cb();
+      });
+    });
+
+    web3LocationPromise.then((_web3Location) => {
+      blockchainConnectorReady = true;
+      this.events.emit('blockchain:connector:ready');
+    });
+
+    let web3Location = await web3LocationPromise;
+    web3Location = normalizePath(web3Location, true);
+
+    await this.registerVar('__Web3', require(web3Location));
+
+    const symlinkLocation = await this.generateSymlink(web3Location);
+
+    let code = `\nconst Web3 = global.__Web3 || require('${symlinkLocation}');`;
+    code += `\nglobal.Web3 = Web3;`;
+
+    let linkedModulePath = path.join(this.modulesPath, 'embarkjs-web3');
+    if (process.platform === 'win32') linkedModulePath = linkedModulePath.replace(/\\/g, '\\\\');
+
+    code += `\n
+      const __embarkWeb3 = require('${linkedModulePath}');
+      EmbarkJS.Blockchain.registerProvider('web3', __embarkWeb3.default || __embarkWeb3);
+      EmbarkJS.Blockchain.setProvider('web3', {});
+    `;
+
+    const configPath = toForwardSlashes(dappPath(this.config.embarkConfig.generationDir, constants.dappArtifacts.dir, constants.dappArtifacts.blockchain));
+
+    code += `\nif (!global.__Web3) {`; // Only connect when in the Dapp
+    code += `\n  const web3ConnectionConfig = require('${configPath}');`;
+    code += `\n  EmbarkJS.Blockchain.connect(web3ConnectionConfig, (err) => {if (err) { console.error(err); } });`;
+    code += `\n}`;
+
+    this.embark.addCodeToEmbarkJS(code);
+
+    code = "EmbarkJS.Blockchain.setProvider('web3', {web3});";
+
+    const shouldInit = (_config) => {
+      return true;
+    };
```

Comment:
Not sure how much value it adds to define this function here, over just passing it down as a lambda

Reply:
The code here was ported straight from https://github.com/embark-framework/embark/blob/master/packages/embarkjs-connector-web3/index.js.